openreferee_server/notify.py

```diff
@@ -0,0 +1,45 @@
+import json
+import threading
+
+from .operations import setup_requests_session
+
+
+class NotifyService:
+    def __init__(self, url=None, token=None) -> None:
+        self.url = url
+        self.token = token
+        self.session = None
+
+    def send(self, payload):
+        try:
+            self.session.get(self.url, data=json.dumps({"payload": payload}))
+        except Exception:
+            pass
+
+    def notify(self, payload):
+        if self.url is None:
+            return
+
+        if self.session is None:
+            self.session = setup_requests_session(self.token)
+
+        threading.Thread(target=self.send, args=(payload,)).start()
```

Comment:
I wonder... is all this extra complexity really needed? To me it sounds like the whole code in here could be done with a simple

```python
def notify(payload):
    sess = ...
    try:
        sess.post(..., timeout=5)
    except ...:
        pass  # log it
```

no threads, no flask extension, no context etc needed ;)

Reply:
Sure I thought I might be over-complicating this, but I thought I'd be adding the notify to each of the endpoints, or maybe extending the logic out if required.

The reason why I use an extension, I needed to load in the environmental config for the notify URL. So I figured I'd need an extension to load this in? Should I just load from process.env from anywhere?

The threading is there as I wanted this to me fire and forget so there would be no slow down to this service if our downstream system was offline or slow to response.

The two reasons I'm using a session is that I am able to config the authentication at a parent level using the method you already have, and depending on the traffic in this service, having a common session between requests reduces http overhead. I'm not sure how it cleans itself up though, so I'm happy to simply this if you do not believe there is much traffic?

Reply:
> The reason why I use an extension, I needed to load in the environmental config for the notify URL. So I figured I'd need an extension to load this in? Should I just load from process.env from anywhere?

Yes, you can read env vars anywhere. Or even put it in the `app.config` dict and then access it from there later.

> The threading is there as I wanted this to me fire and forget so there would be no slow down to this service if our downstream system was offline or slow to response.

Fair enough, I don't know how reliable your downstream is. Personally I'd probably just use a low-enough timeout and then only log request-related errors so they don't break the flow.

> The two reasons I'm using a session is that I am able to config the authentication at a parent level using the method you already have, and depending on the traffic in this service, having a common session between requests reduces http overhead. I'm not sure how it cleans itself up though, so I'm happy to simply this if you do not believe there is much traffic?

Actually, I'm not sure if the session actually keeps connections alive etc. or not. Using it for the token is a good idea though!

Reply:
Is this call to the microservice blocking, as in, will buttons take longer to load if we put in a downstream call? 

I expect that the (yet-to-be-built) downstream service may be hosted on the other side of the planet, also I really didn't want to be responsible for an extra 2-5 second load times on some buttons, or prematurely timeout cause of distance.

That being said, I'm also no python guru, so maybe threads are not the best tool for the job here (due to leaky threads or contention issues)? So if you feel strongly about this I will just change it. 

Also regarding the use of an extension, I was just following a pattern I'm familiar with from other (non-python) frameworks, so if you want me to simplify it, no problem, I'm just not sure if you comment was more of an observation or a suggestion for change.

Reply:
the buttons won't take longer to load since you're not doing anything to the `get_custom_revision_actions` endpoint, but still it's probably better to do it this way if you're unsure of the reliability of the service